### PR TITLE
culling & lod support

### PR DESCRIPTION
**Pull request by arose**
- [Breaking] Add culling & LOD support
    - Cull per-object and per-instance
    - Cull based on frustum and camera distance
    - LOD visibility based on camera distance
    - Special LOD mode for spheres with automatic levels
    - Occlusion culling (only WebGL2)
        - Hi-Z pass
        - Cull based on previous frame's Hi-Z buffer
- Add stochstic/dithered transparency to fade overlapping LODs in and out
- Add "Automatic Detail" preset that shows surface/cartoon/ball & stick based on camera distance

### CODE REVIEW

**Comment by giagitom** [comment]
@arose Is there an example that allows to observe the effects of these new options?

**Comment by arose** [comment]
@giagitom for now https://molstar.org/dev/me/examples/ is the best example. For a technical overview you can watch https://www.khronos.org/events/webgl-webgpu-meetup-november-2023.

**Comment by giagitom** [comment]
Awesome, Thanks

**Comment by arose** [comment]
Also added fading LODs in/out using stochastic transparency. You can easily try it with the "Automatic Detail" preset

image

File: src/mol-canvas3d/passes/hi-z.ts [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+        for (let i = 0, il = this.levelData.length; i < il; ++i) {
+            const td = this.levelData[i];
+            td.framebuffer.bind();
+
+            if (i > 0) {
+                ValueCell.update(v.uInvSize, td.invSize);
+                ValueCell.update(v.uOffset, Vec2.set(v.uOffset.ref.value, 0, 0));
+                ValueCell.update(v.tPreviousLevel, this.levelData[i - 1].texture);
+            } else {
+                const s = Math.pow(2, Math.ceil(Math.log(Math.max(gl.drawingBufferWidth, gl.drawingBufferHeight)) / Math.log(2)) - 1);
```
Comment (dsehnal) [inline src/mol-canvas3d/passes/hi-z.ts:196]: This is independent of the for loop, move it up?

File: src/mol-canvas3d/passes/hi-z.ts [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+            position: 'absolute',
+            pointerEvents: 'none',
+        });
+        element.parentElement.appendChild(rect);
+
+        this.debug = { container, canvas, ctx, rect };
+    }
+
+    private showRect(p: Vec4, occluded: boolean) {
+        if (!this.supported || !this.props.enabled || !this.ready || !this.debug) return;
```
Comment (dsehnal) [inline src/mol-canvas3d/passes/hi-z.ts:433]: make this a single property since it's repeated quite often, maybe `get canDebug()` ?

**Review by dsehnal — APPROVED** [review]
nice 👍

File: src/mol-canvas3d/passes/hi-z.ts (hunk 1 above)
Comment (arose) [reply src/mol-canvas3d/passes/hi-z.ts:196]: thanks, done

File: src/mol-canvas3d/passes/hi-z.ts (hunk 2 above)
Comment (arose) [reply src/mol-canvas3d/passes/hi-z.ts:433]: why not :)

**Pull request merged**